Author: Luca Boccassi <luca.boccassi@gmail.com>
Description: Redefine WRITE_COMBINE_FLUSH on ARM on kernel >= 4.3
 From kernel 4.3 and newer (commit f81309067ff2) ARM function outer_sync is
 no longer present. Instead, behind a CONFIG_ARM_HEAVY_MB flag, arm_heavy_mb
 is defined. Fall back to no memory barrier at all if that is not configured.

--- a/nv-linux.h
+++ b/nv-linux.h
@@ -457,7 +457,13 @@ typedef spinlock_t                nv_spi
 #if defined(CONFIG_OUTER_CACHE)
 #define OUTER_FLUSH_RANGE(start, end)   outer_flush_range((start),(end))
 #endif
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,3,0)
 #define WRITE_COMBINE_FLUSH()    { dsb(); outer_sync(); }
+#elif defined(CONFIG_ARM_HEAVY_MB)
+#define WRITE_COMBINE_FLUSH()    { dsb(); arm_heavy_mb(); }
+#else
+#define WRITE_COMBINE_FLUSH()    { dsb(); }
+#endif
 #elif defined(NVCPU_AARCH64)
     static inline void nv_flush_cache_cpu(void *info)
     {
